src/java/org/apache/cassandra/index/sai/disk/v1/trie/InvertedIndexWriter.java

```diff
@@ -0,0 +1,112 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.index.sai.disk.v1.trie;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import javax.annotation.concurrent.NotThreadSafe;
+
+import org.apache.commons.lang3.mutable.MutableLong;
+
+import org.apache.cassandra.index.sai.IndexContext;
+import org.apache.cassandra.index.sai.disk.PostingList;
+import org.apache.cassandra.index.sai.disk.TermsIterator;
+import org.apache.cassandra.index.sai.disk.format.IndexComponent;
+import org.apache.cassandra.index.sai.disk.format.IndexDescriptor;
+import org.apache.cassandra.index.sai.disk.v1.SegmentMetadata;
+import org.apache.cassandra.index.sai.disk.v1.postings.PostingsWriter;
+import org.apache.cassandra.index.sai.utils.SAICodecUtils;
+import org.apache.cassandra.utils.bytecomparable.ByteComparable;
+
+/**
+ * Builds an on-disk inverted index structure: terms dictionary and postings lists.
+ */
+@NotThreadSafe
+public class InvertedIndexWriter implements Closeable
+{
+    private final TrieTermsDictionaryWriter termsDictionaryWriter;
+    private final PostingsWriter postingsWriter;
+    private long postingsAdded;
+
+    public InvertedIndexWriter(IndexDescriptor indexDescriptor, IndexContext indexContext, boolean segmented) throws IOException
+    {
+        this.termsDictionaryWriter = new TrieTermsDictionaryWriter(indexDescriptor, indexContext, segmented);
+        this.postingsWriter = new PostingsWriter(indexDescriptor, indexContext, segmented);
+    }
+
+    /**
+     * Appends a set of terms and associated postings to their respective overall SSTable component files.
+     *
+     * @param terms an iterator of terms with their associated postings
+     *
+     * @return metadata describing the location of this inverted index in the overall SSTable
+     *         terms and postings component files
+     */
+    public SegmentMetadata.ComponentMetadataMap writeAll(TermsIterator terms) throws IOException
+    {
+        // Terms and postings writers are opened in append mode with pointers at the end of their respective files.
+        long termsOffset = termsDictionaryWriter.getStartOffset();
+        long postingsOffset = postingsWriter.getStartOffset();
+
+        while (terms.hasNext())
+        {
+            ByteComparable term = terms.next();
+            try (PostingList postings = terms.postings())
+            {
+                final long offset = postingsWriter.write(postings);
+                if (offset >= 0)
```

Comment:
nit: Maybe we should have a specific constant shared here and in `PostingsWriter`, which seems to return -1 if there are no postings, and zero isn't even possible?

Reply:
To be honest, this doesn't make much sense. The `PostingsWriter` specifically checks that the `postings.size() > 0` so we can't possibly write no postings. In which case the `offset` doesn't need checking here at all. We can assume that it is a valid offset.

I would be more inclined to put an assert in the `PostingsWriter` to make sure that postings have been written.

Reply:
> I would be more inclined to put an assert in the PostingsWriter

SGTM